swarm/network/stream/stream.go

```diff
@@ -0,0 +1,441 @@
+	live      bool
+	stream    string
+	key       []byte
+	next      chan error
+}
+
+// Client interface for incoming peer Streamer
+type Client interface {
+	NeedData([]byte) func()
+	BatchDone(string, uint64, []byte, []byte) func() (*TakeoverProof, error)
+	Close()
+}
+
+// nextBatch adjusts the indexes by inspecting the intervals
+func (c *client) nextBatch(from uint64) (nextFrom uint64, nextTo uint64) {
+	var intervals []uint64
+	if c.live {
+		if len(intervals) == 0 {
+			intervals = []uint64{c.sessionAt, from}
+		} else {
+			intervals[1] = from
+		}
+		nextFrom = from
+	} else if from >= c.sessionAt { // history sync complete
+		intervals = nil
+		nextFrom = from
+		nextTo = math.MaxUint64
+	} else if len(intervals) > 2 && from >= intervals[2] { // filled a gap in the intervals
+		intervals = append(intervals[:1], intervals[3:]...)
+		nextFrom = intervals[1]
+		if len(intervals) > 2 {
+			nextTo = intervals[2]
+		} else {
+			nextTo = c.sessionAt
+		}
+	} else {
+		nextFrom = from
+		intervals[1] = from
+		nextTo = c.sessionAt
+	}
+	// b.intervals.set(intervals)
+	return nextFrom, nextTo
+}
+
+func (c *client) batchDone(p *Peer, req *OfferedHashesMsg, hashes []byte) error {
+	if tf := c.BatchDone(req.Stream, req.From, hashes, req.Root); tf != nil {
+		tp, err := tf()
+		if err != nil {
+			return err
+		}
+		return p.SendPriority(tp, c.priority)
+	}
+	return nil
+}
+
+func (c *client) close() {
+	close(c.next)
+	c.Close()
+}
+
+// Spec is the spec of the streamer protocol
+var Spec = &protocols.Spec{
+	Name:       "stream",
+	Version:    1,
+	MaxMsgSize: 10 * 1024 * 1024,
+	Messages: []interface{}{
+		UnsubscribeMsg{},
+		OfferedHashesMsg{},
+		WantedHashesMsg{},
+		TakeoverProofMsg{},
+		SubscribeMsg{},
+		RetrieveRequestMsg{},
+		ChunkDeliveryMsg{},
+		SubscribeErrorMsg{},
+	},
+}
+
+func (r *Registry) Protocols() []p2p.Protocol {
+	return []p2p.Protocol{
+		{
+			Name:    Spec.Name,
+			Version: Spec.Version,
+			Length:  Spec.Length(),
+			Run:     r.runProtocol,
+			// NodeInfo: ,
+			// PeerInfo: ,
+		},
+	}
+}
+
+func (r *Registry) APIs() []rpc.API {
+	return []rpc.API{
+		{
+			Namespace: "stream",
+			Version:   "0.1",
+			Service:   r.api,
+			Public:    true,
+		},
+	}
+}
+
+func (r *Registry) Start(server *p2p.Server) error {
+	r.api.dpa.Start()
+	return nil
+}
+
+func (r *Registry) Stop() error {
+	r.api.dpa.Stop()
+	return nil
+}
+
+type API struct {
+	streamer *Registry
+	dpa      *storage.DPA
+}
+
+func NewAPI(r *Registry, store storage.ChunkStore) *API {
+	dpa := storage.NewDPA(store, storage.NewChunkerParams())
+	return &API{
+		streamer: r,
+		dpa:      dpa,
+	}
+}
+
+func readAll(dpa *storage.DPA, hash []byte) (int64, error) {
+	r := dpa.Retrieve(hash)
+	buf := make([]byte, 1024)
+	var n int
+	var total int64
+	var err error
+	for (total == 0 || n > 0) && err == nil {
+		n, err = r.ReadAt(buf, total)
+		total += int64(n)
+	}
+	if err != nil && err != io.EOF {
+		return total, err
+	}
+	return total, nil
```

Comment:
If the idea is to read all the data and return the number of bytes read, perhaps use `io.Copy(ioutil.Discard, r)`?